# SPRINT 1: IMDB Ratings Project Revised


# Introduction
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

# Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Data Overview

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/datasets/movies_and_shows.csv')

print(df.info())
print()
print(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB
None

              name                Character   r0le        TITLE   Type  \
0   Robert De Niro            Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster            Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                      Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  Tax

# Conclusion
We can seet that there are TV shows that were released before 1999, so we will have to remove this data from the dataframe.

## Data Preprocessing

### Format column headers

In [3]:
# List the names of the columns
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

In [4]:
# Rename columns
df = df.rename(columns = {
    '   name': 'name',
    'Character': 'character', 
    'r0le': 'role', 
    'TITLE': 'title', 
    '  Type': 'type', 
    'release Year': 'release_year',
    'imdb sc0re': 'imdb_score', 
    'imdb v0tes': 'imdb_votes'
})

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          85579 non-null  object 
 1   character     85579 non-null  object 
 2   role          85579 non-null  object 
 3   title         85578 non-null  object 
 4   type          85579 non-null  object 
 5   release_year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb_score    80970 non-null  float64
 8   imdb_votes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


### Missing values

In [5]:
# Calculating number of missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

In [6]:
# Checking the percentage of missing values
number = df.isnull().sum()/len(df)*100
rounded = round(number, 0)
print(rounded)

name            0.0
character       0.0
role            0.0
title           0.0
type            0.0
release_year    0.0
genres          0.0
imdb_score      5.0
imdb_votes      6.0
dtype: float64


# Conclusion
Not all missing values affect the research: the single missing title is not critical. The missing values in column 'imdb_score' and 'imdb_votes' represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potential affect our results. To avoid this issue, we will drop all rows with missing values in 'imdb_score' and 'imdb_votes' columns.

In [7]:
# Dropping all rows where columns with title, scores, and votes have missing values
df = df.dropna(subset=['title', 'imdb_score', 'imdb_votes'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80853 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          80853 non-null  object 
 1   character     80853 non-null  object 
 2   role          80853 non-null  object 
 3   title         80853 non-null  object 
 4   type          80853 non-null  object 
 5   release_year  80853 non-null  int64  
 6   genres        80853 non-null  object 
 7   imdb_score    80853 non-null  float64
 8   imdb_votes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 6.2+ MB


In [8]:
# Check for any missing values and count them
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates

In [9]:
# Count the number of duplicate rows
df.duplicated().sum()

6994

Review the duplicated rows to determine if removing them would distort our dataset

In [10]:
# Produce table with duplicates
df_duplicates = df[df.duplicated()]
df_duplicates.tail()

name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

In [11]:
# Removing duplicated rows
df = df.drop_duplicates()

In [12]:
# Check for duplicates and count
df.duplicated().sum()

0

Now get rid of implicit duplicates in the 'type' column. For example, the string 'SHOW' can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique 'type' names, sorted in alphabetical order: To do so:

    - Retrieve the intended dataframe column
    - Apply a sorting method to it
    - For the sorted column, call the method that will return all unique column values

In [13]:
# Viewing unique type names
df['type'].unique()

array(['MOVIE', 'the movie', 'tv show', 'shows', 'movies', 'tv shows',
       'tv series', 'tv', 'SHOW'], dtype=object)

In [14]:
# Function to correct the names in the 'type' columns
def replace_wrong_show(wrong_show_list, correct_show):
    df['type'] = df['type'].replace(wrong_show_list, correct_show)
    return df

Call replace_wrong_show() and pass it the arguments so that it clears implicit duplicates and replaces them with 'SHOW':

In [15]:
# Removing implicit duplicates
df = replace_wrong_show(df['type'].unique(), 'SHOW')

In [16]:
# Check for implicit duplicates in the 'type' column
df['type'].unique()

array(['SHOW'], dtype=object)

# Conclusion
We detected three issues with the data:

    1. Incorrect header styles
    2. Missing values
    3. Duplicated rows and implicit duplicates
    
The headers have been cleaned up to make preprocessing the table simplier.

All rows with missing values have been removed.

The absence of duplicates will make the results more precis and easier to understand.

Now, we can move on to our analysis of the prepared data.

## Data Analysis 

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [17]:
# Filter the data so it contains shows released after 1998
df_filtered = df[df['release_year'] > 1998]

df_filtered.head()

name            character   role     title  type  \
1664       Jeff Probst       Himself - Host  ACTOR  Survivor  SHOW   
1955  Benicio del Toro  Franky Four Fingers  ACTOR    Snatch  SHOW   
1956     Dennis Farina           Cousin Avi  ACTOR    Snatch  SHOW   
1957      Vinnie Jones    Bullet Tooth Tony  ACTOR    Snatch  SHOW   
1958         Brad Pitt        Mickey O'Neil  ACTOR    Snatch  SHOW   

      release_year               genres  imdb_score  imdb_votes  
1664          2000          ['reality']         7.4     24687.0  
1955          2000  ['crime', 'comedy']         8.3    841435.0  
1956          2000  ['crime', 'comedy']         8.3    841435.0  
1957          2000  ['crime', 'comedy']         8.3    841435.0  
1958          2000  ['crime', 'comedy']         8.3    841435.0

In [18]:
# Filter the data so only the SHOWS are in the database
df_filtered = df_filtered[df_filtered['type']=='SHOW']

df_filtered.head()

name            character   role     title  type  \
1664       Jeff Probst       Himself - Host  ACTOR  Survivor  SHOW   
1955  Benicio del Toro  Franky Four Fingers  ACTOR    Snatch  SHOW   
1956     Dennis Farina           Cousin Avi  ACTOR    Snatch  SHOW   
1957      Vinnie Jones    Bullet Tooth Tony  ACTOR    Snatch  SHOW   
1958         Brad Pitt        Mickey O'Neil  ACTOR    Snatch  SHOW   

      release_year               genres  imdb_score  imdb_votes  
1664          2000          ['reality']         7.4     24687.0  
1955          2000  ['crime', 'comedy']         8.3    841435.0  
1956          2000  ['crime', 'comedy']         8.3    841435.0  
1957          2000  ['crime', 'comedy']         8.3    841435.0  
1958          2000  ['crime', 'comedy']         8.3    841435.0

In [24]:
# Round imdb score to nearest whole number
df_filtered['imdb_score'] = df_filtered['imdb_score'].round()

df_filtered.head()

name            character   role     title  type  \
1664       Jeff Probst       Himself - Host  ACTOR  Survivor  SHOW   
1955  Benicio del Toro  Franky Four Fingers  ACTOR    Snatch  SHOW   
1956     Dennis Farina           Cousin Avi  ACTOR    Snatch  SHOW   
1957      Vinnie Jones    Bullet Tooth Tony  ACTOR    Snatch  SHOW   
1958         Brad Pitt        Mickey O'Neil  ACTOR    Snatch  SHOW   

      release_year               genres  imdb_score  imdb_votes  
1664          2000          ['reality']         7.0     24687.0  
1955          2000  ['crime', 'comedy']         8.0    841435.0  
1956          2000  ['crime', 'comedy']         8.0    841435.0  
1957          2000  ['crime', 'comedy']         8.0    841435.0  
1958          2000  ['crime', 'comedy']         8.0    841435.0

It's time to identify outliers based on the number of votes.

In [21]:
# Use groupby() for scores and count all unique values in each group, print the result
unique_count = df_filtered.groupby('imdb_score').size()

print(unique_count)

imdb_score
2.0       207
3.0       562
4.0      3169
5.0      8926
6.0     23201
7.0     21517
8.0     11568
9.0       723
10.0        8
dtype: int64


Based on the aggregation performed, it is evident that scores 2 (207 voted shows), 3 (562 voted shows), 9 (723 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-8 as acceptable), use conditional filtering and grouping.

In [26]:
# Filter dataframe using two conditions (scores to be in the range 4-8)
df_filtered = df_filtered[(df_filtered['imdb_score'] > 3) & (df_filtered['imdb_score'] < 9)]

In [27]:
# Group scores and corresponding average number of votes, reset index and print the result
result = df_filtered.groupby('imdb_score')['imdb_votes'].mean().reset_index()

In [31]:
df_filtered.head()

name            character   role     title  type  \
1664       Jeff Probst       Himself - Host  ACTOR  Survivor  SHOW   
1955  Benicio del Toro  Franky Four Fingers  ACTOR    Snatch  SHOW   
1956     Dennis Farina           Cousin Avi  ACTOR    Snatch  SHOW   
1957      Vinnie Jones    Bullet Tooth Tony  ACTOR    Snatch  SHOW   
1958         Brad Pitt        Mickey O'Neil  ACTOR    Snatch  SHOW   

      release_year               genres  imdb_score  imdb_votes  
1664          2000          ['reality']         7.0     24687.0  
1955          2000  ['crime', 'comedy']         8.0    841435.0  
1956          2000  ['crime', 'comedy']         8.0    841435.0  
1957          2000  ['crime', 'comedy']         8.0    841435.0  
1958          2000  ['crime', 'comedy']         8.0    841435.0

In [32]:
result.head()

imdb_score     imdb_votes
0         4.0    8699.529820
1         5.0   14236.204123
2         6.0   28807.336192
3         7.0   51308.138542
4         8.0  145532.023081

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [34]:
# Rename columns
result.columns = ['imdb_score', 'AverageVotes']

# Round to nearest whole number
result['AverageVotes'] = result['AverageVotes'].round()

# Print dataframe in descending order
result = result.sort_values(by='AverageVotes', ascending=False)

print(result)

   imdb_score  AverageVotes
4         8.0      145532.0
3         7.0       51308.0
2         6.0       28807.0
1         5.0       14236.0
0         4.0        8700.0


The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

# Final Conclusion with Step-by-Step Overview


### ✅ Final Conclusion

This analysis set out to investigate a key question:  
**Do the highest-rated TV shows from the "Golden Age" of television (1999–present) also receive the most votes on IMDb?**

To answer this, we followed a structured data analysis process. Below is an overview of the steps and our findings:

---

### 🔍 Step-by-Step Overview

1. **Data Overview**
   - Loaded the dataset and examined structure and content.
   - Identified missing values, duplicates, and formatting issues.

2. **Data Preprocessing**
   - Cleaned column headers and removed rows with missing or unreliable data.
   - Eliminated duplicate and inconsistent entries, especially in the `type` column.
   - Filtered the dataset to include only **TV shows** released from **1999 onward**.

3. **Data Analysis**
   - Grouped IMDb scores into whole numbers (e.g., 7.4 → 7) for clarity.
   - Excluded outlier scores with insufficient vote data (scores 1, 2, 3, 9, and 10 had too little votes).
   - Calculated the **average number of votes** per IMDb score.
   - Sorted results to determine which scores had the highest engagement.

---

### 📈 Key Finding

- TV shows with **higher IMDb scores (7 and 8)** have the **highest average number of votes**.
- This pattern confirms the assumption that **highly-rated shows also receive more audience engagement**.

---

### 📌 Business Insight

This insight can support content investment decisions by highlighting that:
- Viewers are more likely to engage (via voting) with higher-quality content.
- IMDb ratings and vote counts can serve as reliable proxies for show popularity and reach.

As a result, studios, platforms, and marketers should prioritize promoting or acquiring **top-rated content**, especially those rated **7 and above**, to maximize audience interest and platform engagement.
""")
